src/cmd.rs

```diff
@@ -70,30 +71,30 @@ impl<'a, T: FromRedisValue> Iterator for Iter<'a, T> {
 #[cfg(feature = "aio")]
 use crate::aio::ConnectionLike as AsyncConnection;
 
-/// Represents a redis iterator that can be used with async connections.
+/// The inner future of AsyncIter
 #[cfg(feature = "aio")]
-pub struct AsyncIter<'a, T: FromRedisValue + 'a> {
+struct AsyncIterInner<'a, T: FromRedisValue + 'a> {
     batch: std::vec::IntoIter<T>,
     con: &'a mut (dyn AsyncConnection + Send + 'a),
     cmd: Cmd,
 }
 
+/// Represents the state of AsyncIter
 #[cfg(feature = "aio")]
-impl<'a, T: FromRedisValue + 'a> AsyncIter<'a, T> {
-    /// ```rust,no_run
-    /// # use redis::AsyncCommands;
-    /// # async fn scan_set() -> redis::RedisResult<()> {
-    /// # let client = redis::Client::open("redis://127.0.0.1/")?;
-    /// # let mut con = client.get_async_connection().await?;
-    /// con.sadd("my_set", 42i32).await?;
-    /// con.sadd("my_set", 43i32).await?;
-    /// let mut iter: redis::AsyncIter<i32> = con.sscan("my_set").await?;
-    /// while let Some(element) = iter.next_item().await {
-    ///     assert!(element == 42 || element == 43);
-    /// }
-    /// # Ok(())
-    /// # }
-    /// ```
+enum IterOrFuture<'a, T: FromRedisValue + 'a> {
+    Iter(AsyncIterInner<'a, T>),
+    Future(BoxFuture<'a, (AsyncIterInner<'a, T>, Option<T>)>),
```

Comment:
It might be better to use the future directly (instead of BoxFuture) together with e.g. `pin-project-lite`.

Reply:
I'll take a look at pin-project, since I never used it before, what is the advantage of that, not having a heap allocation?

Reply:
Here it would be mostly to avoid an allocation and the dynamic dispatch.

However, I just realized that this future is coming from an async block/fn, so I don't think it's possible to avoid this without TAIT (or a bigger code change), better keep it this way then, sorry for the noise.

Reply:
Don't worry, I learn about TAIT thanks to your comment :)